src/bytearray.rs

```diff
@@ -0,0 +1,236 @@
+    fn borrow(&self) -> &[u8; N] {
+        &self.bytes
+    }
+}
+impl<const N: usize> BorrowMut<[u8; N]> for ByteArray<N> {
+    fn borrow_mut(&mut self) -> &mut [u8; N] {
+        &mut self.bytes
+    }
+}
+
+impl<const N: usize> Deref for ByteArray<N> {
+    type Target = [u8; N];
+
+    fn deref(&self) -> &Self::Target {
+        &self.bytes
+    }
+}
+
+impl<const N: usize> DerefMut for ByteArray<N> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.bytes
+    }
+}
+
+impl<const N: usize> Borrow<Bytes> for ByteArray<N> {
+    fn borrow(&self) -> &Bytes {
+        Bytes::new(&self.bytes)
+    }
+}
+
+impl<const N: usize> BorrowMut<Bytes> for ByteArray<N> {
+    fn borrow_mut(&mut self) -> &mut Bytes {
+        unsafe { &mut *(&mut self.bytes as &mut [u8] as *mut [u8] as *mut Bytes) }
+    }
+}
+
+impl<Rhs, const N: usize> PartialEq<Rhs> for ByteArray<N>
+where
+    Rhs: ?Sized + Borrow<[u8; N]>,
+{
+    fn eq(&self, other: &Rhs) -> bool {
+        self.as_ref().eq(other.borrow())
+    }
+}
+
+impl<Rhs, const N: usize> PartialOrd<Rhs> for ByteArray<N>
+where
+    Rhs: ?Sized + Borrow<[u8; N]>,
+{
+    fn partial_cmp(&self, other: &Rhs) -> Option<Ordering> {
+        self.as_ref().partial_cmp(other.borrow())
+    }
+}
+
+impl<const N: usize> Hash for ByteArray<N> {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.bytes.hash(state);
+    }
+}
+
+impl<const N: usize> IntoIterator for ByteArray<N> {
+    type Item = u8;
+    type IntoIter = <[u8; N] as IntoIterator>::IntoIter;
+
+    fn into_iter(self) -> Self::IntoIter {
+        IntoIterator::into_iter(self.bytes)
+    }
+}
+
+impl<'a, const N: usize> IntoIterator for &'a ByteArray<N> {
+    type Item = &'a u8;
+    type IntoIter = <&'a [u8; N] as IntoIterator>::IntoIter;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.bytes.iter()
+    }
+}
+
+impl<'a, const N: usize> IntoIterator for &'a mut ByteArray<N> {
+    type Item = &'a mut u8;
+    type IntoIter = <&'a mut [u8; N] as IntoIterator>::IntoIter;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.bytes.iter_mut()
+    }
+}
+
+impl<const N: usize> Serialize for ByteArray<N> {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        serializer.serialize_bytes(&self.bytes)
+    }
+}
+
+struct ByteArrayVisitor<const N: usize>;
+
+impl<'de, const N: usize> Visitor<'de> for ByteArrayVisitor<N> {
+    type Value = ByteArray<N>;
+
+    fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+        write!(formatter, "[u8; {}]", N)
+    }
+
+    fn visit_seq<V>(self, mut visitor: V) -> Result<ByteArray<N>, V::Error>
+    where
+        V: SeqAccess<'de>,
+    {
+        let mut bytes = [0; N];
+        let mut idx = 0;
+        while let Some(b) = visitor.next_element()? {
+            bytes[idx] = b;
+            idx += 1;
+            if idx == N {
+                break;
+            }
+        }
```

Comment:
Recommend using the array to drive the loop here, instead of a manual loop counter. Additionally, I believe that the argument to `invalid_length` is supposed to be the number of elements from the data source, rather than the expected correct length.

```
let mut bytes = [0; N];

for (idx, byte) in bytes.iter_mut().enumerate() {
    *byte = visitor
        .next_element()?
        .ok_or_else(|| V::Error::invalid_length(idx, &self)?;
}

Ok(ByteArray::from(bytes))
```

Reply:
Done in [167c622](https://github.com/serde-rs/bytes/pull/28/commits/167c62287f7fd808faa8187f003d9a62026badd7)